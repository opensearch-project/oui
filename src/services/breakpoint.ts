/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { keysOf } from '../components/common';

export type OuiBreakpointSize = 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl' | 'xxxl';

export type OuiBreakpoints = {
  /**
   * Set the minimum window width at which to start to the breakpoint
   */
  [key in OuiBreakpointSize]: number;
};

export const BREAKPOINTS: OuiBreakpoints = {
  xxxl: 1920,
  xxl: 1680,
  xl: 1200,
  l: 992,
  m: 768,
  s: 575,
  xs: 0,
};

export const BREAKPOINT_KEYS = keysOf(BREAKPOINTS);

/**
 * Given the current `width` and an object of `OuiBreakpoints`,
 * this function returns the string that is the name of the breakpoint key
 * that is less than or equal to the width
 *
 * @param {number} width Can either be the full window width or any width
 * @param {OuiBreakpoints} breakpoints An object with keys for sizing and values for minimum width
 * @param {OuiBreakpointSize[] | undefined} allowedBreakpoints An array of allowed breakpoints
 * @returns {string | undefined} Name of the breakpoint key or `undefined` if a key doesn't exist
 */
export function getBreakpoint(
  width: number,
  breakpoints: OuiBreakpoints = BREAKPOINTS,
  allowedBreakpoints: OuiBreakpointSize[] = keysOf(BREAKPOINTS)
): OuiBreakpointSize | undefined {
  // Find the breakpoint (key) whose value is <= windowWidth starting with the largest first
  return keysOf(BREAKPOINTS).find(
    (key) => allowedBreakpoints.includes(key) && breakpoints[key] <= width
  );
}

/**
 * Given the current `width` and a max breakpoint key,
 * this function returns true or false if the `width` falls within the max
 * breakpoint or any breakpoints below
 *
 * @param {number} width Can either be the full window width or any width
 * @param {OuiBreakpointSize | number} max The named breakpoint or custom number to check against
 * @param {OuiBreakpoints} breakpoints An object with keys for sizing and values for minimum width
 * @returns {boolean} Will return `false` if it can't find a value for the `max` breakpoint
 */
export function isWithinMaxBreakpoint(
  width: number,
  max: OuiBreakpointSize | number,
  breakpoints: OuiBreakpoints = BREAKPOINTS
): boolean {
  if (typeof max === 'number') {
    return width <= max;
  } else {
    const currentBreakpoint = getBreakpoint(width, breakpoints);
    return currentBreakpoint
      ? breakpoints[currentBreakpoint] <= breakpoints[max]
      : false;
  }
}

/**
 * Given the current `width` and a max breakpoint key,
 * this function returns true or false if the `width` falls within the max
 * breakpoint or any breakpoints below
 *
 * @param {number} width Can either be the full window width or any width
 * @param {OuiBreakpointSize | number} min The named breakpoint or custom number to check against
 * @param {OuiBreakpoints} breakpoints An object with keys for sizing and values for minimum width
 * @returns {boolean} Will return `false` if it can't find a value for the `min` breakpoint
 */
export function isWithinMinBreakpoint(
  width: number,
  min: OuiBreakpointSize | number,
  breakpoints: OuiBreakpoints = BREAKPOINTS
): boolean {
  if (typeof min === 'number') {
    return width >= min;
  } else {
    const currentBreakpoint = getBreakpoint(width, breakpoints);
    return currentBreakpoint
      ? breakpoints[currentBreakpoint] >= breakpoints[min]
      : false;
  }
}

/**
 * Given the current `width` and an array of breakpoint keys,
 * this function returns true or false if the `width` falls within
 * any of the named breakpoints
 *
 * @param {number} width Can either be the full window width or any width
 * @param {OuiBreakpointSize[]} sizes An array of named breakpoints
 * @param {OuiBreakpoints} breakpoints An object with keys for sizing and values for minimum width
 * @returns {boolean} Returns `true` if current breakpoint name is included in `sizes`
 */
export function isWithinBreakpoints(
  width: number,
  sizes: OuiBreakpointSize[],
  breakpoints: OuiBreakpoints = BREAKPOINTS
): boolean {
  const currentBreakpoint = getBreakpoint(width, breakpoints);
  return currentBreakpoint ? sizes.includes(currentBreakpoint) : false;
}

/* OUI -> EUI Aliases */
export type EuiBreakpointSize = OuiBreakpointSize;
export type EuiBreakpoints = OuiBreakpoints;
/* End of Aliases */
