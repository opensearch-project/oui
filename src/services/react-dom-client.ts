/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import * as ReactDOM from 'react-dom';

/**
 * Root interface for React 16 compatibility with React 18's createRoot API
 */
export interface Root {
  render(element: React.ReactElement): void;
  unmount(): void;
}

/**
 * Shim for React 18's createRoot API
 * This provides a compatible API for React 16 using ReactDOM.render
 *
 * @param container The DOM element to render into
 * @returns A root object with render and unmount methods
 */
export function createRoot(container: Element): Root {
  return {
    render(element: React.ReactElement) {
      ReactDOM.render(element, container);
    },
    unmount() {
      ReactDOM.unmountComponentAtNode(container);
    },
  };
}
