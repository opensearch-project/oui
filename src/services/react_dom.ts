/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
// eslint-disable-next-line import/no-unresolved
import { createRoot } from './react-dom-client';

/**
 * Version-agnostic render function that works with both React 16 and 18
 * In React 16, it uses our shim for react-dom/client (./react-dom-client.ts)
 * In React 18, we can alias this file to the actual react-dom/client
 */
export function render(element: React.ReactElement, container: Element) {
  const root = createRoot(container);
  root.render(element);
  return root;
}
