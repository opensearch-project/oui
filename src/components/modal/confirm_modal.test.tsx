/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';

import { requiredProps } from '../../test';
import { keys } from '../../services';

import {
  CANCEL_BUTTON,
  CONFIRM_BUTTON,
  OuiConfirmModal,
} from './confirm_modal';

let onConfirm: jest.Mock;
let onCancel: jest.Mock;

beforeEach(() => {
  onConfirm = jest.fn();
  onCancel = jest.fn();
});

describe('OuiConfirmModal', () => {
  test('renders OuiConfirmModal', () => {
    render(
      <OuiConfirmModal
        title="A confirmation modal"
        onCancel={() => {}}
        onConfirm={onConfirm}
        cancelButtonText="Cancel Button Text"
        confirmButtonText="Confirm Button Text"
        {...requiredProps}>
        This is a confirmation modal example
      </OuiConfirmModal>
    );
    expect(document.body).toMatchSnapshot();
  });

  test('renders OuiConfirmModal without OuiModalBody, if empty', () => {
    render(
      <OuiConfirmModal
        title="A confirmation modal"
        onCancel={() => {}}
        onConfirm={onConfirm}
        cancelButtonText="Cancel Button Text"
        confirmButtonText="Confirm Button Text"
        {...requiredProps}
      />
    );
    expect(document.body).toMatchSnapshot();
  });

  test('onConfirm', () => {
    const { getByTestId } = render(
      <OuiConfirmModal
        onCancel={onCancel}
        onConfirm={onConfirm}
        cancelButtonText="Cancel Button Text"
        confirmButtonText="Confirm Button Text"
      />
    );

    fireEvent.click(getByTestId('confirmModalConfirmButton'));
    expect(onConfirm).toHaveBeenCalledTimes(1);
    expect(onCancel).toHaveBeenCalledTimes(0);
  });

  test('isLoading', () => {
    const { getByTestId } = render(
      <OuiConfirmModal
        onCancel={onCancel}
        onConfirm={onConfirm}
        isLoading
        cancelButtonText="Cancel Button Text"
        confirmButtonText="Confirm Button Text"
      />
    );

    fireEvent.click(getByTestId('confirmModalConfirmButton'));
    expect(onConfirm).toHaveBeenCalledTimes(0);
  });

  test('onConfirm can be disabled', () => {
    const { getByTestId } = render(
      <OuiConfirmModal
        onCancel={onCancel}
        onConfirm={onConfirm}
        cancelButtonText="Cancel Button Text"
        confirmButtonText="Confirm Button Text"
        confirmButtonDisabled={true}
      />
    );

    fireEvent.click(getByTestId('confirmModalConfirmButton'));
    expect(onConfirm).toHaveBeenCalledTimes(0);
    expect(onCancel).toHaveBeenCalledTimes(0);
  });

  describe('onCancel', () => {
    test('triggerd by click', () => {
      const { getByTestId } = render(
        <OuiConfirmModal
          onCancel={onCancel}
          onConfirm={onConfirm}
          cancelButtonText="Cancel Button Text"
          confirmButtonText="Confirm Button Text"
        />
      );

      fireEvent.click(getByTestId('confirmModalCancelButton'));
      expect(onConfirm).toHaveBeenCalledTimes(0);
      expect(onCancel).toHaveBeenCalledTimes(1);
    });

    test('triggered by esc key', () => {
      const { getByTestId } = render(
        <OuiConfirmModal
          onCancel={onCancel}
          onConfirm={onConfirm}
          cancelButtonText="Cancel Button Text"
          confirmButtonText="Confirm Button Text"
          data-test-subj="modal"
        />
      );

      fireEvent.keyDown(getByTestId('modal'), {
        key: keys.ESCAPE,
      });
      expect(onConfirm).toHaveBeenCalledTimes(0);
      expect(onCancel).toHaveBeenCalledTimes(1);
    });
  });

  describe('defaultFocusedButton', () => {
    test('is cancel', async () => {
      const { getByTestId } = render(
        <OuiConfirmModal
          onCancel={onCancel}
          onConfirm={onConfirm}
          cancelButtonText="Cancel Button Text"
          confirmButtonText="Confirm Button Text"
          defaultFocusedButton={CANCEL_BUTTON}
        />
      );

      // The auto-focus implementation waits a frame before focusing.
      await waitFor(() => {
        const button = getByTestId('confirmModalCancelButton');
        expect(document.activeElement).toEqual(button);
      });
    });

    test('is confirm', async () => {
      const { getByTestId } = render(
        <OuiConfirmModal
          onCancel={onCancel}
          onConfirm={onConfirm}
          cancelButtonText="Cancel Button Text"
          confirmButtonText="Confirm Button Text"
          defaultFocusedButton={CONFIRM_BUTTON}
        />
      );

      // The auto-focus implementation waits a frame before focusing.
      await waitFor(() => {
        const button = getByTestId('confirmModalConfirmButton');
        expect(document.activeElement).toEqual(button);
      });
    });
  });
});
