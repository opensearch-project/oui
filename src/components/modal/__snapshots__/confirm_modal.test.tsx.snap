// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiConfirmModal renders OuiConfirmModal 1`] = `
Array [
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />,
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="1"
  />,
  <div
    data-focus-lock-disabled="false"
  >
    <div
      aria-label="aria-label"
      class="ouiModal ouiModal--maxWidth-default ouiModal--confirmation testClass1 testClass2"
      data-test-subj="test subject string"
      tabindex="0"
    >
      <button
        aria-label="Closes this modal window"
        class="ouiButtonIcon ouiButtonIcon--text ouiButtonIcon--empty ouiButtonIcon--xSmall ouiModal__closeIcon"
        type="button"
      >
        <span
          aria-hidden="true"
          class="ouiButtonIcon__icon"
          color="inherit"
          data-ouiicon-type="cross"
        />
      </button>
      <div
        class="ouiModal__flex"
      >
        <div
          class="ouiModalHeader"
        >
          <div
            class="ouiModalHeader__title"
            data-test-subj="confirmModalTitleText"
          >
            A confirmation modal
          </div>
        </div>
        <div
          class="ouiModalBody"
        >
          <div
            class="ouiModalBody__overflow"
          >
            <div
              class="ouiText ouiText--medium"
              data-test-subj="confirmModalBodyText"
            >
              <p>
                This is a confirmation modal example
              </p>
            </div>
          </div>
        </div>
        <div
          class="ouiModalFooter"
        >
          <button
            class="ouiButtonEmpty ouiButtonEmpty--primary"
            data-test-subj="confirmModalCancelButton"
            type="button"
          >
            <span
              class="ouiButtonContent ouiButtonEmpty__content"
            >
              <span
                class="ouiButtonEmpty__text"
              >
                Cancel Button Text
              </span>
            </span>
          </button>
          <button
            class="ouiButton ouiButton--primary ouiButton--fill"
            data-test-subj="confirmModalConfirmButton"
            type="button"
          >
            <span
              class="ouiButtonContent ouiButton__content"
            >
              <span
                class="ouiButton__text"
              >
                Confirm Button Text
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />,
]
`;

exports[`OuiConfirmModal renders OuiConfirmModal without OuiModalBody, if empty 1`] = `
Array [
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />,
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="1"
  />,
  <div
    data-focus-lock-disabled="false"
  >
    <div
      aria-label="aria-label"
      class="ouiModal ouiModal--maxWidth-default ouiModal--confirmation testClass1 testClass2"
      data-test-subj="test subject string"
      tabindex="0"
    >
      <button
        aria-label="Closes this modal window"
        class="ouiButtonIcon ouiButtonIcon--text ouiButtonIcon--empty ouiButtonIcon--xSmall ouiModal__closeIcon"
        type="button"
      >
        <span
          aria-hidden="true"
          class="ouiButtonIcon__icon"
          color="inherit"
          data-ouiicon-type="cross"
        />
      </button>
      <div
        class="ouiModal__flex"
      >
        <div
          class="ouiModalHeader"
        >
          <div
            class="ouiModalHeader__title"
            data-test-subj="confirmModalTitleText"
          >
            A confirmation modal
          </div>
        </div>
        <div
          class="ouiModalFooter"
        >
          <button
            class="ouiButtonEmpty ouiButtonEmpty--primary"
            data-test-subj="confirmModalCancelButton"
            type="button"
          >
            <span
              class="ouiButtonContent ouiButtonEmpty__content"
            >
              <span
                class="ouiButtonEmpty__text"
              >
                Cancel Button Text
              </span>
            </span>
          </button>
          <button
            class="ouiButton ouiButton--primary ouiButton--fill"
            data-test-subj="confirmModalConfirmButton"
            type="button"
          >
            <span
              class="ouiButtonContent ouiButton__content"
            >
              <span
                class="ouiButton__text"
              >
                Confirm Button Text
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />,
]
`;
